File: src/shared/gpu/protocol.ts
import { GPU_PREVIEW_SIZE, GPU_WORKING_RES_MAX } from './topNodes'

export { GPU_PREVIEW_SIZE, GPU_WORKING_RES_MAX }

/** Length-prefixed IPC magic: 'PFGP'. */
export const GPU_IPC_MAGIC = 0x50464750
export const GPU_PROTOCOL_VERSION = 1

export interface GpuHelloOk {
  kind: 'hello-ok'
  gpu: boolean
  share: 'syphon' | 'spout' | 'none'
  error: string | null
}

export interface GpuCompileNode {
  id: string
  type: string
  width: number
  height: number
  /** Pixel-port name → upstream node id. */
  inputs: Record<string, string>
}

export interface GpuCompileRequest {
  nodes: GpuCompileNode[]
  pixelCount: number
  resolutionWidth: number
  resolutionHeight: number
}

export interface GpuGradientStop {
  t: number
  r: number
  g: number
  b: number
}

export interface GpuNodeUniforms {
  floats: number[]
  colours: number[]
  ints: number[]
  strings: string[]
  stops?: GpuGradientStop[]
}

export interface GpuShareIn {
  nodeId: string
  sender: string
}

export interface GpuShareOut {
  nodeId: string
  name: string
  width: number
  height: number
  mapping: 'grid' | 'layout'
  sourceNodeId: string
  fromCpu: boolean
}

export interface GpuMediaRef {
  nodeId: string
  path: string
  kind: 'video' | 'image'
}

export interface GpuFrameRequest {
  timeMs: number
  deltaMs: number
  liveNodeIds: string[]
  uniforms: Record<string, GpuNodeUniforms>
  cpuUploadIds: string[]
  sampleNodeIds: string[]
  previewNodeIds: string[]
  feedbackResets: string[]
  media: GpuMediaRef[]
  shareIn: GpuShareIn[]
  shareOut: GpuShareOut[]
}

export interface GpuBakeRequest {
  durationMs: number
  fps: number
  sampleNodeId: string
}

export type GpuRequest =
  | { id: number; kind: 'hello' }
  | { id: number; kind: 'compile'; body: GpuCompileRequest }
  | { id: number; kind: 'frame'; body: GpuFrameRequest }
  | { id: number; kind: 'bake'; body: GpuBakeRequest }
  | { id: number; kind: 'shutdown' }

export interface GpuPreviewBlob {
  nodeId: string
  width: number
  height: number
}

export interface GpuFrameResultBody {
  error: string | null
  shareSenders: string[]
  shareError: string | null
  sampleIds: string[]
  previews: GpuPreviewBlob[]
}

export interface GpuBakeResultBody {
  error: string | null
  frameCount: number
  pixelCount: number
  fps: number
}

export type GpuResponse =
  | { id: number; kind: 'hello-ok'; body: GpuHelloOk }
  | { id: number; kind: 'compile-ok' }
  | { id: number; kind: 'compile-error'; error: string }
  | { id: number; kind: 'frame-ok'; body: GpuFrameResultBody }
  | { id: number; kind: 'bake-ok'; body: GpuBakeResultBody }
  | { id: number; kind: 'error'; error: string }

File: gpu-engine/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Deserialize)]
pub struct Request {
    pub id: u32,
    pub kind: String,
    #[serde(default)]
    pub body: Value,
}

#[derive(Debug, Serialize)]
#[allow(dead_code)]
pub struct Response {
    pub id: u32,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<Value>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompileNode {
    pub id: String,
    #[serde(rename = "type")]
    pub ty: String,
    pub width: u32,
    pub height: u32,
    pub inputs: std::collections::HashMap<String, String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompileRequest {
    pub nodes: Vec<CompileNode>,
    pub pixel_count: u32,
    pub resolution_width: u32,
    pub resolution_height: u32,
}

#[derive(Debug, Deserialize, Clone)]
pub struct GradientStop {
    pub t: f32,
    pub r: f32,
    pub g: f32,
    pub b: f32,
}

#[derive(Debug, Deserialize, Clone, Default)]
pub struct NodeUniforms {
    #[serde(default)]
    pub floats: Vec<f32>,
    #[serde(default)]
    pub colours: Vec<f32>,
    #[serde(default)]
    pub ints: Vec<i32>,
    #[serde(default)]
    pub strings: Vec<String>,
    #[serde(default)]
    pub stops: Option<Vec<GradientStop>>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaRef {
    pub node_id: String,
    pub path: String,
    pub kind: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShareIn {
    pub node_id: String,
    pub sender: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShareOut {
    pub node_id: String,
    pub name: String,
    pub width: u32,
    pub height: u32,
    pub mapping: String,
    pub source_node_id: String,
    pub from_cpu: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CpuUpload {
    pub node_id: String,
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FrameRequest {
    pub time_ms: f32,
    pub delta_ms: f32,
    pub live_node_ids: Vec<String>,
    pub uniforms: std::collections::HashMap<String, NodeUniforms>,
    #[serde(default)]
    pub cpu_uploads: Vec<CpuUpload>,
    pub sample_node_ids: Vec<String>,
    #[serde(default)]
    pub preview_node_ids: Vec<String>,
    #[serde(default)]
    pub feedback_resets: Vec<String>,
    #[serde(default)]
    pub media: Vec<MediaRef>,
    #[serde(default)]
    pub share_in: Vec<ShareIn>,
    #[serde(default)]
    pub share_out: Vec<ShareOut>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BakeRequest {
    pub duration_ms: f32,
    pub fps: f32,
    pub sample_node_id: String,
}
